actions/src/sema4ai/actions/agent/__init__.py

```
@@ -8,9 +8,75 @@
+
+
+# Cache PyArrow availability to avoid repeated imports
+_pyarrow_available = None
+
+def _is_pyarrow_available() -> bool:
+    """Check if PyArrow is available, caching the result."""
+    global _pyarrow_available
+    if _pyarrow_available is None:
+        try:
+            import pyarrow as pa
+            import pyarrow.parquet as pq
+            _pyarrow_available = True
+        except ImportError:
+            _pyarrow_available = False
+    return _pyarrow_available
+
+
+def _parse_dataframe_response(response) -> dict:
+    """Parse dataframe response, trying Parquet first, then falling back to JSON.
+    
+    Args:
+        response: HTTP response object
+        
+    Returns:
+        dict: Parsed dataframe data with columns, rows, name, description
+    """
+    import json
+    
+    # Check if response is Parquet format and PyArrow is available
+    content_type = response.headers.get('content-type', '').lower()
+    if ('parquet' in content_type or 'application/octet-stream' in content_type) and _is_pyarrow_available():
+        try:
+            import pyarrow as pa
+            import pyarrow.parquet as pq
+            from io import BytesIO
+            
+            # Read Parquet data
+            parquet_data = BytesIO(response.content)
+            table = pq.read_table(parquet_data)
+            
+            # Convert to our format
+            columns = table.column_names
+            rows = []
+            for i in range(len(table)):
+                row = []
+                for col in columns:
+                    value = table[col][i].as_py()
+                    row.append(value)
+                rows.append(row)
+            
+            return {
+                "columns": columns,
+                "rows": rows,
+                "name": None,  # Parquet doesn't include metadata
+                "description": None
+            }
+        except Exception:
+            # Parquet parsing failed, fall back to JSON
```

I think you should just let the exception be raised here (if we requested in parquet format, it'd be surprising to have the response as json).